server/src/instant/lib/ring/websocket.clj

```diff
@@ -21,43 +22,71 @@
     BufferedBinaryMessage
     BufferedTextMessage
     CloseMessage
-    WebSocketChannel
+    StreamSourceFrameChannel WebSocketChannel
     WebSockets
     WebSocketCallback]
    [io.undertow.websockets.spi WebSocketHttpExchange]
    [org.xnio ChannelListener]
    [ring.adapter.undertow Util]
    [clojure.lang IPersistentMap]
    [io.undertow.websockets.extensions PerMessageDeflateHandshake]
-   [java.util.concurrent.locks ReentrantLock]))
+   [java.util.concurrent.locks ReentrantLock]
+   [java.util.concurrent.atomic AtomicLong]
+   [java.nio ByteBuffer]
+   [org.xnio IoUtils]))
 
 (defn ws-listener
   "Creates an `AbstractReceiveListener`. This relays calls to 
    `on-message`, `on-close-message`, and `on-error` callbacks. 
    
    See `ws-callback` for more details."
-  [{:keys [on-message on-close-message on-error channel-wrapper]}]
+  [{:keys [on-message on-close-message on-error channel-wrapper atomic-last-received-at]}]
   (let [on-message       (or on-message (constantly nil))
         on-error         (or on-error (constantly nil))
         on-close-message (or on-close-message (constantly nil))]
     (proxy [AbstractReceiveListener] []
       (onFullTextMessage [^WebSocketChannel channel ^BufferedTextMessage message]
+
+        (.set atomic-last-received-at (System/currentTimeMillis))
         (on-message {:channel (channel-wrapper channel)
                      :data    (.getData message)}))
       (onFullBinaryMessage [^WebSocketChannel channel ^BufferedBinaryMessage message]
+        (.set atomic-last-received-at (System/currentTimeMillis))
         (let [pooled (.getData message)]
           (try
             (let [payload (.getResource pooled)]
               (on-message {:channel (channel-wrapper channel)
                            :data    (Util/toArray payload)}))
             (finally (.free pooled)))))
+      (onPong [^WebSocketChannel channel ^StreamSourceFrameChannel channel]
```

Comment:
You may be wondering: 

How does the client know to send a `pong`, we did not change Reactor. 

Most clients now can handle websocket pings by default. So since the server sends `ping`, chrome will automatically respond with a pong. 

(Tested this with chrome and expo iOS)